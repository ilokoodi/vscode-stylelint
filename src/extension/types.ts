// eslint-disable-next-line node/no-unpublished-import
import type TypedEventEmitter from 'typed-emitter';

/**
 * VS Code extension event names.
 */
export interface ExtensionEvents {
	DidRegisterDocumentFormattingEditProvider: () => void;
}

/**
 * VS Code extension public API.
 */
export type PublicApi = TypedEventEmitter<ExtensionEvents> & {
	formattingReady: boolean;
};

/**
 * Extension API event names.
 */
export enum ApiEvent {
	DidRegisterDocumentFormattingEditProvider = 'DidRegisterDocumentFormattingEditProvider',
}
