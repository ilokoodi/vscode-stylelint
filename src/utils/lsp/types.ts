import type LSP from 'vscode-languageserver-protocol';

/**
 * Code actions for a particular rule.
 */
export interface RuleCodeActions {
	/**
	 * Action to disable the rule for a specific line.
	 */
	disableLine?: LSP.CodeAction;

	/**
	 * Action to disable the rule for an entire file.
	 */
	disableFile?: LSP.CodeAction;

	/**
	 * Action to show documentation for the rule.
	 */
	documentation?: LSP.CodeAction;
}
