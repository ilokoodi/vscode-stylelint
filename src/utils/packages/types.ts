import { Stylelint } from '../stylelint/index';

/**
 * Package manager identifiers.
 */
export type PackageManager = 'npm' | 'yarn' | 'pnpm';

/**
 * Options for resolving the Stylelint package.
 */
export type ResolverOptions = {
	packageManager?: PackageManager;
	stylelintPath?: string;
};

/**
 * Stylelint package resolution result.
 */
export type StylelintResolutionResult = {
	stylelint: Stylelint;
	resolvedPath: string;
};

/**
 * A tracer function that can be used to log messages.
 */
export type TracerFn = (message: string, verbose?: string) => void;
