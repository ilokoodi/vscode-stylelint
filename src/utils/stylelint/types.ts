import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line node/no-unpublished-import
import type stylelint from 'stylelint';
import type { PackageManager } from '../packages';

/**
 * Diagnostics for a lint run.
 */
export type LintDiagnostics = {
	/**
	 * The diagnostics, each corresponding to a warning or error emitted by
	 * Stylelint.
	 */
	diagnostics: LSP.Diagnostic[];

	/**
	 * Raw output from Stylelint, if any.
	 */
	output?: string;
};

/**
 * Disable report rule names.
 */
export enum DisableReportRuleNames {
	Needless = '--report-needless-disables',
	InvalidScope = '--report-invalid-scope-disables',
	Descriptionless = '--report-descriptionless-disables',
	Illegal = 'reportDisables',
}

/**
 * Stylelint runner options.
 */
export type RunnerOptions = {
	config?: stylelint.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager?: PackageManager;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet?: string[];
	stylelintPath?: string;
	validate?: string[];
};

/**
 * Error thrown when a rule's option is invalid.
 */
export class InvalidOptionError extends Error {
	reasons: string[];

	constructor(warnings: { text: string }[]) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}

// TODO: Create type upstream
// https://github.com/stylelint/stylelint/pull/5696
/**
 * A Stylelint configuration error. Taken from
 * https://github.com/stylelint/stylelint/blob/551dcb5/lib/utils/configurationError.js
 */
export type ConfigurationError = Error & { code: 78 };
