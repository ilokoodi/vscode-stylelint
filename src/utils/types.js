'use strict';

const { CodeActionKind: VSCodeActionKind } = require('vscode-languageserver-types');

/**
 * Command IDs
 * @enum {string}
 */
const CommandId = {
	ApplyAutoFix: 'stylelint.applyAutoFix',
};

/**
 * Code action kinds
 * @enum {string}
 */
const CodeActionKind = {
	StylelintSourceFixAll: `${VSCodeActionKind.SourceFixAll}.stylelint`,
};

/**
 * Disable report rule names
 * @enum {string}
 */
const DisableReportRuleNames = {
	Needless: '--report-needless-disables',
	InvalidScope: '--report-invalid-scope-disables',
	Descriptionless: '--report-descriptionless-disables',
	Illegal: 'reportDisables',
};

/**
 * Error thrown when a rule's option is invalid.
 */
class InvalidOptionError extends Error {
	/** @param {{text: string}[]} warnings */
	constructor(warnings) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}

module.exports = {
	CommandId,
	CodeActionKind,
	DisableReportRuleNames,
	InvalidOptionError,
};
