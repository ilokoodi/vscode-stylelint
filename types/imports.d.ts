import * as postcss from 'postcss';
export import postcss = postcss;

import * as stylelint from 'stylelint';
export import stylelint = stylelint;

import * as vscode from 'vscode';
export import vscode = vscode;

import * as vscodeLanguageServer from 'vscode-languageserver';
export import vscodeLanguageServer = vscodeLanguageServer;

import * as vscodeLanguageServerTextDocument from 'vscode-languageserver-textdocument';
export import vscodeLanguageServerTextDocument = vscodeLanguageServerTextDocument;

import * as vscodeLanguageServerTypes from 'vscode-languageserver-types';
export import vscodeLanguageServerTypes = vscodeLanguageServerTypes;

export as namespace imports;
