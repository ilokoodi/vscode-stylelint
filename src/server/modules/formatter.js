'use strict';

const { DocumentFormattingRequest } = require('vscode-languageserver-protocol');
const { formattingOptionsToRules } = require('../../utils/stylelint');
const { Notification } = require('../../utils/types');

/**
 * @implements {LanguageServerModule}
 */
class FormatterModule {
	static id = 'formatter';

	/**
	 * The language server context.
	 * @type {LanguageServerContext}
	 */
	#context;

	/**
	 * The logger to use, if any.
	 * @type {winston.Logger | undefined}
	 */
	#logger;

	/**
	 * Whether or not dynamic registration for document formatting should be
	 * attempted.
	 * @type {boolean}
	 */
	#registerDynamically = false;

	/**
	 * The disposable for the dynamically registered document formatter.
	 * @type {lsp.Disposable | undefined}
	 */
	#registration = undefined;

	/**
	 * @param {LanguageServerModuleConstructorParameters} params
	 */
	constructor({ context, logger }) {
		this.#context = context;
		this.#logger = logger;
	}

	/**
	 * @param {lsp.TextDocument} document
	 * @returns {boolean}
	 */
	#shouldFormat(document) {
		return this.#context.options.validate.includes(document.languageId);
	}

	/**
	 * @param {lsp.InitializeParams} params
	 * @returns {Partial<lsp.InitializeResult>}
	 */
	onInitialize({ capabilities }) {
		this.#registerDynamically = Boolean(capabilities.textDocument?.formatting?.dynamicRegistration);

		return {
			capabilities: {
				// Use static registration if dynamic registration is not
				// supported by the client
				documentFormattingProvider: !this.#registerDynamically,
			},
		};
	}

	/**
	 * @returns {void}
	 */
	onDidRegisterHandlers() {
		this.#logger?.debug('Registering onDocumentFormatting handler');

		this.#context.connection.onDocumentFormatting(({ textDocument, options }) => {
			this.#logger?.debug('Received onDocumentFormatting', { textDocument, options });

			if (!textDocument) {
				this.#logger?.debug('No text document provided, ignoring');

				return null;
			}

			const { uri } = textDocument;
			const document = this.#context.documents.get(uri);

			if (!document || !this.#shouldFormat(document)) {
				if (this.#logger?.isDebugEnabled()) {
					if (!document) {
						this.#logger.debug('Unknown document, ignoring', { uri });
					} else {
						this.#logger.debug('Document should not be formatted, ignoring', {
							uri,
							language: document.languageId,
						});
					}
				}

				return null;
			}

			const linterOptions = {
				config: {
					rules: formattingOptionsToRules(options),
				},
			};

			this.#logger?.debug('Formatting document', { uri, linterOptions });

			const fixes = this.#context.getFixes(document, linterOptions);

			this.#logger?.debug('Returning fixes', { uri, fixes });

			return fixes;
		});

		this.#logger?.debug('onDocumentFormatting handler registered');
	}

	/**
	 * @param {DidChangeValidateLanguagesParams} params
	 * @returns {Promise<void>}
	 */
	async onDidChangeValidateLanguages({ languages }) {
		if (this.#logger?.isDebugEnabled()) {
			this.#logger?.debug('Received onDidChangeValidateLanguages', { languages: [...languages] });
		}

		// If dynamic registration is supported and the list of languages that should be validated
		// has changed, then (re-)register the formatter.
		if (this.#registerDynamically) {
			// Dispose the old formatter registration if it exists.
			if (this.#registration) {
				this.#logger?.debug('Disposing old formatter registration');

				this.#registration.dispose();

				this.#logger?.debug('Old formatter registration disposed');
			}

			// If there are languages that should be validated, register a formatter for those
			// languages.
			if (languages.size > 0) {
				const documentSelector = [];

				for (const language of languages) {
					documentSelector.push({ language });
				}

				if (this.#logger?.isDebugEnabled()) {
					this.#logger?.debug('Registering formatter for languages', { languages: [...languages] });
				}

				this.#registration = await this.#context.connection.client.register(
					DocumentFormattingRequest.type,
					{ documentSelector },
				);

				this.#context.connection.sendNotification(
					Notification.DidRegisterDocumentFormattingEditProvider,
					{},
				);

				this.#logger?.debug('Formatter registered');
			}
		}
	}
}

module.exports = {
	FormatterModule,
};
