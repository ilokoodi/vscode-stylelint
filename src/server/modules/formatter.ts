import { DocumentFormattingRequest } from 'vscode-languageserver-protocol';
import type { TextDocument } from 'vscode-languageserver-textdocument';
import type LSP from 'vscode-languageserver-protocol';
import { formattingOptionsToRules } from '../../utils/stylelint';
import { Notification } from '../types';
import type {
	DidChangeValidateLanguagesParams,
	LanguageServerContext,
	LanguageServerModuleConstructorParameters,
	LanguageServerModule,
} from '../types';
import type winston from 'winston';

export class FormatterModule implements LanguageServerModule {
	static id = 'formatter';

	/**
	 * The language server context.
	 */
	#context: LanguageServerContext;

	/**
	 * The logger to use, if any.
	 */
	#logger: winston.Logger | undefined;

	/**
	 * Whether or not dynamic registration for document formatting should be
	 * attempted.
	 */
	#registerDynamically = false;

	/**
	 * The disposable for the dynamically registered document formatter.
	 */
	#registration: LSP.Disposable | undefined = undefined;

	constructor({ context, logger }: LanguageServerModuleConstructorParameters) {
		this.#context = context;
		this.#logger = logger;
	}

	#shouldFormat(document: TextDocument): boolean {
		return this.#context.options.validate.includes(document.languageId);
	}

	onInitialize({ capabilities }: LSP.InitializeParams): Partial<LSP.InitializeResult> {
		this.#registerDynamically = Boolean(capabilities.textDocument?.formatting?.dynamicRegistration);

		return {
			capabilities: {
				// Use static registration if dynamic registration is not
				// supported by the client
				documentFormattingProvider: !this.#registerDynamically,
			},
		};
	}

	onDidRegisterHandlers(): void {
		this.#logger?.debug('Registering onDocumentFormatting handler');

		this.#context.connection.onDocumentFormatting(({ textDocument, options }) => {
			this.#logger?.debug('Received onDocumentFormatting', { textDocument, options });

			if (!textDocument) {
				this.#logger?.debug('No text document provided, ignoring');

				return null;
			}

			const { uri } = textDocument;
			const document = this.#context.documents.get(uri);

			if (!document || !this.#shouldFormat(document)) {
				if (this.#logger?.isDebugEnabled()) {
					if (!document) {
						this.#logger.debug('Unknown document, ignoring', { uri });
					} else {
						this.#logger.debug('Document should not be formatted, ignoring', {
							uri,
							language: document.languageId,
						});
					}
				}

				return null;
			}

			const linterOptions = {
				config: {
					rules: formattingOptionsToRules(options),
				},
			};

			this.#logger?.debug('Formatting document', { uri, linterOptions });

			const fixes = this.#context.getFixes(document, linterOptions);

			this.#logger?.debug('Returning fixes', { uri, fixes });

			return fixes;
		});

		this.#logger?.debug('onDocumentFormatting handler registered');
	}

	async onDidChangeValidateLanguages({
		languages,
	}: DidChangeValidateLanguagesParams): Promise<void> {
		if (this.#logger?.isDebugEnabled()) {
			this.#logger?.debug('Received onDidChangeValidateLanguages', { languages: [...languages] });
		}

		// If dynamic registration is supported and the list of languages that should be validated
		// has changed, then (re-)register the formatter.
		if (this.#registerDynamically) {
			// Dispose the old formatter registration if it exists.
			if (this.#registration) {
				this.#logger?.debug('Disposing old formatter registration');

				this.#registration.dispose();

				this.#logger?.debug('Old formatter registration disposed');
			}

			// If there are languages that should be validated, register a formatter for those
			// languages.
			if (languages.size > 0) {
				const documentSelector = [];

				for (const language of languages) {
					documentSelector.push({ language });
				}

				if (this.#logger?.isDebugEnabled()) {
					this.#logger?.debug('Registering formatter for languages', { languages: [...languages] });
				}

				this.#registration = await this.#context.connection.client.register(
					DocumentFormattingRequest.type,
					{ documentSelector },
				);

				this.#context.connection.sendNotification(
					Notification.DidRegisterDocumentFormattingEditProvider,
					{},
				);

				this.#logger?.debug('Formatter registered');
			}
		}
	}
}
