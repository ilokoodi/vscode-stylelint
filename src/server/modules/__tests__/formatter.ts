jest.mock('vscode-uri', () => {
	const vscodeUri = jest.requireActual('vscode-uri');
	const mock = jest.createMockFromModule<typeof vscodeUri>('vscode-uri');

	mock.URI.parse = (uri: string) => {
		const parsed = vscodeUri.URI.parse(uri);

		if (parsed.scheme === 'file') {
			return parsed;
		}

		const fsPath = parsed.fsPath?.replace(/\//g, '\\');

		return {
			...parsed,
			fsPath,
		};
	};

	return mock;
});

import * as LSP from 'vscode-languageserver-protocol';
import { Position, TextEdit } from 'vscode-languageserver-types';
import type winston from 'winston';
import type { LanguageServerOptions, LanguageServerModuleConstructorParameters } from '../../types';

import { Notification } from '../../types';
import { FormatterModule } from '../formatter';

const mockOptions: LanguageServerOptions = {
	packageManager: 'npm',
	validate: [],
	snippet: [],
};

const mockContext = {
	connection: {
		onDocumentFormatting: jest.fn(),
		onNotification: jest.fn(),
		sendNotification: jest.fn(),
		client: { register: jest.fn() },
	},
	documents: {
		get: jest.fn(),
		onDidOpen: jest.fn(),
		onDidChangeContent: jest.fn(),
		onDidClose: jest.fn(),
		onDidSave: jest.fn(),
	},
	getOptions: jest.fn(async () => mockOptions),
	getFixes: jest.fn(),
};

const mockLogger = {
	debug: jest.fn(),
	isDebugEnabled: jest.fn(() => true),
} as unknown as jest.Mocked<winston.Logger>;

const getParams = (passLogger = false) =>
	({
		context: mockContext,
		logger: passLogger ? mockLogger : undefined,
	} as unknown as LanguageServerModuleConstructorParameters);

describe('FormatterModule', () => {
	beforeEach(() => {
		mockOptions.validate = [];
		jest.clearAllMocks();
	});

	test('should be constructable', () => {
		expect(() => new FormatterModule(getParams())).not.toThrow();
	});

	test('without client dynamic registration support, onInitialize should request static registration', () => {
		const module = new FormatterModule(getParams());

		expect(
			module.onInitialize({
				capabilities: {
					textDocument: {
						formatting: { dynamicRegistration: false },
					},
				},
			} as unknown as LSP.InitializeParams),
		).toMatchSnapshot();
	});

	test('with client dynamic registration support, onInitialize should not request static registration', () => {
		const module = new FormatterModule(getParams());

		expect(
			module.onInitialize({
				capabilities: {
					textDocument: {
						formatting: { dynamicRegistration: true },
					},
				},
			} as unknown as LSP.InitializeParams),
		).toMatchSnapshot();
	});

	test('onDidRegisterHandlers should register a document formatting command handler', () => {
		const module = new FormatterModule(getParams());

		module.onDidRegisterHandlers();

		expect(mockContext.connection.onDocumentFormatting).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.onDocumentFormatting).toHaveBeenCalledWith(expect.any(Function));
	});

	test('should format documents', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
		});
		mockOptions.validate = ['bar'];
		mockContext.getFixes.mockReturnValue([TextEdit.insert(Position.create(0, 0), 'text')]);

		const module = new FormatterModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const result = await handler({
			textDocument: { uri: 'foo' },
			options: {
				insertSpaces: true,
				tabSize: 2,
				insertFinalNewline: true,
				trimFinalNewlines: false,
			},
		});

		expect(result).toMatchSnapshot();
		expect(mockContext.getFixes.mock.calls[0]).toMatchSnapshot();
		expect(mockContext.getFixes.mock.results[0].value).toStrictEqual(result);
	});

	test('with incorrect command, should not attempt to format', async () => {
		const module = new FormatterModule(getParams());

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const result = await handler({ command: 'foo' });

		expect(result).toBeNull();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
	});

	test('with no text document, should not attempt to format', async () => {
		const module = new FormatterModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const result = await handler({ options: {} });

		expect(result).toBeNull();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('No text document provided, ignoring');
	});

	test('if no matching document exists, should not attempt to format', async () => {
		mockContext.documents.get.mockReturnValue(undefined);

		const module = new FormatterModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const result = await handler({
			textDocument: { uri: 'foo' },
			options: {},
		});

		expect(result).toBeNull();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Unknown document, ignoring', { uri: 'foo' });
	});

	test('if document language ID is not in options, should not attempt to format', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
		});
		mockOptions.validate = ['baz'];

		const module = new FormatterModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const result = await handler({
			textDocument: { uri: 'foo' },
			options: {},
		});

		expect(result).toBeNull();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document should not be formatted, ignoring',
			{
				uri: 'foo',
				language: 'bar',
			},
		);
	});

	test('with no debug log level and no valid document, should not attempt to log reason', async () => {
		mockContext.documents.get.mockReturnValue({
			uri: 'foo',
			languageId: 'bar',
		});
		mockOptions.validate = ['baz'];
		mockLogger.isDebugEnabled.mockReturnValue(false);

		const module = new FormatterModule(getParams(true));

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onDocumentFormatting.mock.calls[0][0];

		const handlerParams = {
			textDocument: { uri: 'foo' },
			options: {},
		};

		const result = await handler(handlerParams);

		expect(result).toBeNull();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Received onDocumentFormatting',
			handlerParams,
		);
	});

	test("without client dynamic registration support, documents.onDidOpen shouldn't register a formatter", async () => {
		mockOptions.validate = ['bar'];

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: false },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({ document: { uri: 'foo', languageId: 'bar' } });

		expect(mockContext.connection.client.register).not.toHaveBeenCalled();
	});

	test("without client dynamic registration support, documents.onDidChangeContent shouldn't register a formatter", async () => {
		mockOptions.validate = ['bar'];

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: false },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidChangeContentHandler = mockContext.documents.onDidChangeContent.mock.calls[0][0];

		await onDidChangeContentHandler({ document: { uri: 'foo', languageId: 'bar' } });

		expect(mockContext.connection.client.register).not.toHaveBeenCalled();
	});

	test("without client dynamic registration support, documents.onDidSave shouldn't register a formatter", async () => {
		mockOptions.validate = ['bar'];

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: false },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidSaveHandler = mockContext.documents.onDidSave.mock.calls[0][0];

		await onDidSaveHandler({ document: { uri: 'foo', languageId: 'bar' } });

		expect(mockContext.connection.client.register).not.toHaveBeenCalled();
	});

	test('with client dynamic registration support, documents.onDidOpen should register a formatter', async () => {
		mockOptions.validate = ['bar'];

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({ document: { uri: 'file:///dir/test.css', languageId: 'bar' } });
		await onDidOpenHandler({ document: { uri: 'scheme:///dir/test.css', languageId: 'bar' } });

		expect(mockContext.connection.client.register).toHaveBeenCalledWith(
			LSP.DocumentFormattingRequest.type,
			{ documentSelector: [{ scheme: 'file', pattern: '/dir/test.css' }] },
		);
		expect(mockContext.connection.client.register).toHaveBeenCalledWith(
			LSP.DocumentFormattingRequest.type,
			{ documentSelector: [{ scheme: 'scheme', pattern: '\\dir\\test.css' }] },
		);
	});

	test('with client dynamic registration support, documents.onDidChangeContent should register a formatter', async () => {
		mockOptions.validate = ['bar'];

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidChangeContentHandler = mockContext.documents.onDidChangeContent.mock.calls[0][0];

		await onDidChangeContentHandler({
			document: { uri: 'file:///dir/test.css', languageId: 'bar' },
		});

		expect(mockContext.connection.client.register).toHaveBeenCalledWith(
			LSP.DocumentFormattingRequest.type,
			{ documentSelector: [{ scheme: 'file', pattern: '/dir/test.css' }] },
		);
	});

	test('with client dynamic registration support, documents.onDidSave should register a formatter', async () => {
		mockOptions.validate = ['bar'];

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidSaveHandler = mockContext.documents.onDidSave.mock.calls[0][0];

		await onDidSaveHandler({ document: { uri: 'file:///dir/test.css', languageId: 'bar' } });

		expect(mockContext.connection.client.register).toHaveBeenCalledWith(
			LSP.DocumentFormattingRequest.type,
			{ documentSelector: [{ scheme: 'file', pattern: '/dir/test.css' }] },
		);
	});

	test('when a formatter was registered, documents.onDidClose should dispose the old registration', async () => {
		mockOptions.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		const mockRegistration = { dispose: jest.fn() };

		mockContext.connection.client.register.mockResolvedValueOnce(mockRegistration);

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];
		const onDidCloseHandler = mockContext.documents.onDidClose.mock.calls[0][0];

		await onDidOpenHandler({ document: { uri: 'file:///dir/test.css', languageId: 'bar' } });
		await onDidCloseHandler({ document: { uri: 'file:///dir/test.css' } });

		expect(mockRegistration.dispose).toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenCalledWith('Deregistering formatter for document', {
			uri: 'file:///dir/test.css',
		});
	});

	test('when a formatter was not registered, documents.onDidClose should not try to dispose a registration', async () => {
		mockOptions.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidCloseHandler = mockContext.documents.onDidClose.mock.calls[0][0];

		await onDidCloseHandler({ document: { uri: 'file:///dir/test.css' } });

		expect(mockLogger.debug).not.toHaveBeenCalledWith('Deregistering formatter for document', {
			uri: 'file:///dir/test.css',
		});
	});

	test('when formatters were registered, DidChangeConfigurationNotification should deregister all registrations', async () => {
		mockOptions.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		const mockRegistration = { dispose: jest.fn() };

		mockContext.connection.client.register.mockResolvedValue(mockRegistration);

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({ document: { uri: 'file:///dir/test-1.css', languageId: 'bar' } });
		await onDidOpenHandler({ document: { uri: 'file:///dir/test-2.css', languageId: 'bar' } });

		const didChangeConfigurationNotificationHandler =
			mockContext.connection.onNotification.mock.calls.find(
				(call) => call[0] === LSP.DidChangeConfigurationNotification.type,
			)[1];

		await didChangeConfigurationNotificationHandler();

		expect(mockRegistration.dispose).toHaveBeenCalledTimes(2);
	});

	test('when formatters were registered, DidChangeWorkspaceFoldersNotification should deregister all registrations', async () => {
		mockOptions.validate = ['bar'];
		mockLogger.isDebugEnabled.mockReturnValue(true);

		const mockRegistration = { dispose: jest.fn() };

		mockContext.connection.client.register.mockResolvedValue(mockRegistration);

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({ document: { uri: 'file:///dir/test-1.css', languageId: 'bar' } });
		await onDidOpenHandler({ document: { uri: 'file:///dir/test-2.css', languageId: 'bar' } });

		const didChangeWorkspaceFoldersNotificationHandler =
			mockContext.connection.onNotification.mock.calls.find(
				(call) => call[0] === LSP.DidChangeWorkspaceFoldersNotification.type,
			)[1];

		await didChangeWorkspaceFoldersNotificationHandler();

		expect(mockRegistration.dispose).toHaveBeenCalledTimes(2);
	});

	test('with client dynamic registration support, only one formatter should be registered per document', async () => {
		mockOptions.validate = ['bar'];

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];
		const onDidSaveHandler = mockContext.documents.onDidSave.mock.calls[0][0];

		await onDidOpenHandler({ document: { uri: 'file:///dir/test.css', languageId: 'bar' } });
		await onDidSaveHandler({ document: { uri: 'file:///dir/test.css', languageId: 'bar' } });

		expect(mockContext.connection.client.register).toHaveBeenCalledTimes(1);
	});

	test('when a formatter is registered, a notification should be sent', async () => {
		mockOptions.validate = ['bar'];

		const module = new FormatterModule(getParams(true));

		module.onInitialize({
			capabilities: {
				textDocument: {
					formatting: { dynamicRegistration: true },
				},
			},
		} as unknown as LSP.InitializeParams);

		module.onDidRegisterHandlers();

		const onDidOpenHandler = mockContext.documents.onDidOpen.mock.calls[0][0];

		await onDidOpenHandler({ document: { uri: 'file:///dir/test.css', languageId: 'bar' } });

		expect(mockContext.connection.sendNotification).toHaveBeenCalledWith(
			Notification.DidRegisterDocumentFormattingEditProvider,
			{
				uri: 'file:///dir/test.css',
				options: {
					documentSelector: [{ scheme: 'file', pattern: '/dir/test.css' }],
				},
			},
		);
	});
});
