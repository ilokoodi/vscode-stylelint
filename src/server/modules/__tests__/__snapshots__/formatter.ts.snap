// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormatterModule should format documents 1`] = `
Array [
  Object {
    "newText": "text",
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`FormatterModule should format documents 2`] = `
Array [
  FullTextDocument {
    "_content": "a {}",
    "_languageId": "bar",
    "_lineOffsets": undefined,
    "_uri": "foo",
    "_version": 1,
  },
  Object {
    "config": Object {
      "rules": Object {
        "indentation": Array [
          2,
        ],
        "no-missing-end-of-source-newline": true,
      },
    },
  },
]
`;

exports[`FormatterModule with client dynamic registration support, onInitialize should not request static registration 1`] = `
Object {
  "capabilities": Object {
    "documentFormattingProvider": false,
  },
}
`;

exports[`FormatterModule without client dynamic registration support, onInitialize should request static registration 1`] = `
Object {
  "capabilities": Object {
    "documentFormattingProvider": true,
  },
}
`;
